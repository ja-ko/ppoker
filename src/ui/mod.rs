use crossterm::event::KeyEvent;
use ratatui::Frame;
use crate::app::{App, AppResult};
use crate::tui::UiPage;

pub(crate) mod voting;
pub(crate) mod log;


pub enum UIAction{
    Continue,
    ChangeView(UiPage),
    Quit,
}
pub trait Page {
    fn render(&mut self, app: &mut App, frame: &mut Frame);
    fn input(&mut self, app: &mut App, event: KeyEvent) -> AppResult<UIAction>;
}