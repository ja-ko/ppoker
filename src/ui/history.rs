use crossterm::event::{KeyCode, KeyEvent};
use ratatui::layout::{Constraint, Direction, Layout};
use ratatui::prelude::*;
use ratatui::widgets::{Cell, Row, Table, TableState};
use ratatui::Frame;

use crate::app::{App, AppResult, HistoryEntry};
use crate::models::GamePhase;
use crate::ui::voting::{format_vote, render_overview, render_own_vote};
use crate::ui::{
    colored_box_style, footer_entries, format_duration, render_box, render_box_colored, Page,
    UIAction, UiPage,
};

pub struct HistoryPage {
    history_state: TableState,
}

impl HistoryPage {
    pub fn new() -> Self {
        Self {
            history_state: TableState::default(),
        }
    }
}

impl Page for HistoryPage {
    fn render(&mut self, app: &mut App, frame: &mut Frame) {
        if self.history_state.selected().is_none() && app.history.len() > 0 {
            self.history_state.select(Some(0));
        }

        let [header, body, footer] = Layout::default()
            .direction(Direction::Vertical)
            .constraints([
                Constraint::Length(3),
                Constraint::Fill(1),
                Constraint::Length(3),
            ])
            .areas(frame.area());

        render_overview(app, header, frame);
        self.render_main(app, body, frame);
        self.render_footer(app, footer, frame);
    }

    fn input(&mut self, _app: &mut App, event: KeyEvent) -> AppResult<UIAction> {
        return Ok(match event.code {
            KeyCode::Esc | KeyCode::Char('q') => UIAction::Quit,
            KeyCode::Char(c) if c == 'v' || c == '-' || c == 'h' || c.is_ascii_digit() => {
                UIAction::ChangeView(UiPage::Voting)
            }
            KeyCode::Down => {
                if let Some(s) = self.history_state.selected() {
                    let mut new_index = s.saturating_add(1);
                    if new_index >= _app.history.len() {
                        new_index = _app.history.len().saturating_sub(1);
                    }
                    self.history_state.select(Some(new_index));
                }
                UIAction::Continue
            }
            KeyCode::Up => {
                if let Some(s) = self.history_state.selected() {
                    self.history_state.select(Some(s.saturating_sub(1)));
                }
                UIAction::Continue
            }
            _ => UIAction::Continue,
        });
    }
}

impl HistoryPage {
    fn render_main(&mut self, app: &mut App, rect: Rect, frame: &mut Frame) {
        let [history, detail] =
            Layout::horizontal([Constraint::Percentage(40), Constraint::Percentage(60)])
                .areas(rect);

        let [vote_summary, players] =
            Layout::vertical([Constraint::Length(9), Constraint::Fill(1)]).areas(detail);

        let current_entry = self.history_state.selected().map(|idx| &app.history[idx]);

        if let Some(current_entry) = current_entry {
            render_own_vote(
                &current_entry.votes,
                current_entry.average,
                GamePhase::Revealed,
                &current_entry.own_vote,
                &current_entry.deck,
                vote_summary,
                frame,
            );

            render_player_list(&current_entry, players, frame);
        }
        self.render_history(app, history, frame);
    }

    fn render_footer(&mut self, app: &mut App, rect: Rect, frame: &mut Frame) {
        let entries = vec!["Vote", "↑", "↓", "Quit"];
        let mut footer = footer_entries(entries);
        if app.has_updates {
            footer = footer.style(Style::new().yellow());
        }
        frame.render_widget(footer, rect);
    }

    fn render_history(&mut self, app: &mut App, rect: Rect, frame: &mut Frame) {
        let inner = render_box("History", rect, frame);

        let rows: Vec<Row> = app
            .history
            .iter()
            .map(|entry| {
                Row::new(vec![
                    Cell::from(Span::raw(entry.round_number.to_string())),
                    Cell::from(Span::raw(format!("{:.1}", entry.average))),
                    Cell::from(Span::raw(format_duration(&entry.length))),
                ])
            })
            .collect();

        let table = Table::new(
            rows,
            [
                Constraint::Length(5),
                Constraint::Length(8),
                Constraint::Fill(1),
            ],
        )
        .column_spacing(4)
        .header(
            Row::new(vec!["Round", "Average", "Duration"])
                .style(Style::new().bold())
                .bottom_margin(1),
        )
        .highlight_symbol("> ")
        .row_highlight_style(Style::new().on_white().black());

        frame.render_stateful_widget(table, inner, &mut self.history_state);
    }
}

fn render_player_list(entry: &HistoryEntry, rect: Rect, frame: &mut Frame) {
    let inner = render_box_colored(
        "Players",
        colored_box_style(GamePhase::Revealed),
        rect,
        frame,
    );
    let mut longest_name = 0;
    let mut players = entry.votes.clone();
    players.sort();
    let rows: Vec<Row> = players
        .iter()
        .map(|p| {
            if p.name.len() > longest_name {
                longest_name = p.name.len();
            }
            Row::new(vec![
                Cell::from(Span::raw(p.name.as_str())),
                Cell::from(format_vote(&p.vote, &entry.own_vote)),
            ])
        })
        .collect();

    let table = Table::new(
        rows,
        [Constraint::Length(longest_name as u16), Constraint::Fill(1)],
    )
    .column_spacing(4)
    .header(
        Row::new(vec!["Name", "Vote"])
            .style(Style::new().bold())
            .bottom_margin(1),
    );

    frame.render_widget(table, inner);
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::app::tests::create_test_app;
    use crate::ui::tests::{send_input, tick};
    use crate::web::client::tests::LocalMockPokerClient;
    use crate::web::client::PokerClient;
    use insta::assert_snapshot;
    use ratatui::backend::TestBackend;

    #[test]
    fn test_render_page() {
        let mut page = HistoryPage::new();
        let mut app = create_test_app(Box::new(LocalMockPokerClient::new("test")));
        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        tick(&mut terminal, &mut page, &mut app);

        assert_snapshot!("Empty history page", terminal.backend());
    }

    #[test]
    fn test_render_page_with_history() {
        let mut page = HistoryPage::new();
        let mut client = LocalMockPokerClient::new("Alice");

        // Add other players
        let bob_id = client.add_user("Bob");
        let charlie_id = client.add_user("Charlie");

        // First round: Everyone votes numbers
        client.vote(Some("5")).unwrap();
        client.user_vote(&bob_id, Some("3"));
        client.user_vote(&charlie_id, Some("8"));
        client.reveal().unwrap();
        client.reset().unwrap();

        // Second round: Mix of numbers and special votes
        client.vote(Some("13")).unwrap();
        client.user_vote(&bob_id, Some("?"));
        client.user_vote(&charlie_id, Some("8"));
        client.reveal().unwrap();
        client.reset().unwrap();

        // Third round: Some abstentions
        client.vote(Some("3")).unwrap();
        client.user_vote(&bob_id, Some("5"));
        // Charlie doesn't vote
        client.reveal().unwrap();
        client.reset().unwrap();

        let mut app = create_test_app(Box::new(client));

        // Render and snapshot the history page
        let mut terminal = Terminal::new(TestBackend::new(90, 30)).unwrap();
        tick(&mut terminal, &mut page, &mut app);

        assert_snapshot!("History page with multiple rounds", terminal.backend());

        send_input(KeyCode::Down, &mut terminal, &mut page, &mut app);
        assert_snapshot!("History page after pressing down", terminal.backend());
    }
}
