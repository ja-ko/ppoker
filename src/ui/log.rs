use crossterm::event::{KeyCode, KeyEvent};
use log::LevelFilter;
use ratatui::Frame;
use ratatui::layout::{Constraint, Layout};
use ratatui::prelude::*;
use ratatui::style::Style;
use ratatui::widgets::{Block, BorderType, Paragraph, Wrap};
use tui_logger::{TuiLoggerLevelOutput, TuiLoggerSmartWidget, TuiWidgetEvent, TuiWidgetState};

use crate::app::{App, AppResult};
use crate::tui::UiPage;
use crate::ui::{Page, UIAction};

pub struct LogPage {
    state: TuiWidgetState,
}

impl LogPage {
    pub fn new() -> Self {
        Self {
            state: TuiWidgetState::default()
                .set_level_for_target("tungstenite::client", LevelFilter::Warn)
                .set_level_for_target("tungstenite::handshake::client", LevelFilter::Warn)
                .set_level_for_target("ppoker::web::ws", LevelFilter::Info)
        }
    }
}

impl Page for LogPage {
    fn render(&mut self, _app: &mut App, frame: &mut Frame) {
        let mut helptexts: Vec<Span> = vec![];
        helptexts.append(&mut help_spans("h", "Toggle target selector"));
        helptexts.append(&mut help_spans("f", "Toggle focus"));
        helptexts.append(&mut help_spans("UP/DOWN", "Navigate"));
        helptexts.append(&mut help_spans("LEFT/RIGHT", "Reduce/increase level"));
        helptexts.append(&mut help_spans("PAGEUP/PAGEDOWN", "Enter Page mode, scroll up/down"));
        helptexts.append(&mut help_spans("ESCAPE", "Exit page mode"));
        helptexts.append(&mut help_spans("SPACE", "Toggle hiding disabled targets"));
        helptexts.append(&mut help_spans("l", "Leave log view"));
        helptexts.append(&mut help_spans("q", "Quit application"));
        helptexts.pop();

        let help_paragraph = Paragraph::new(Line::from(helptexts))
            .wrap(Wrap { trim: true });

        let help_lines = help_paragraph.line_count(frame.size().width.saturating_sub(2)) as u16;

        let [log, help] = Layout::default()
            .direction(Direction::Vertical)
            .constraints([
                Constraint::Fill(1),
                Constraint::Length(help_lines + 2)
            ])
            .areas(frame.size());

        let widget = TuiLoggerSmartWidget::default()
            .style_error(Style::default().red())
            .style_debug(Style::default().cyan())
            .style_warn(Style::default().light_yellow())
            .style_info(Style::default().white())
            .output_separator('|')
            .output_timestamp(Some("%H:%M:%S".to_string()))
            .output_level(Some(TuiLoggerLevelOutput::Long))
            .output_target(true)
            .output_file(false)
            .output_line(false)
            .state(&self.state);

        frame.render_widget(widget, log);

        let block = Block::bordered()
            .title("Help")
            .title_alignment(Alignment::Left)
            .border_type(BorderType::Rounded);
        let help_inner = block.inner(help);

        frame.render_widget(block, help);

        frame.render_widget(help_paragraph, help_inner);
    }

    fn input(&mut self, _app: &mut App, event: KeyEvent) -> AppResult<UIAction> {
        match event.code.into() {
            KeyCode::Char('q') => return Ok(UIAction::Quit),
            KeyCode::Char(' ') => self.state.transition(TuiWidgetEvent::SpaceKey),
            KeyCode::Esc => self.state.transition(TuiWidgetEvent::EscapeKey),
            KeyCode::PageUp => self.state.transition(TuiWidgetEvent::PrevPageKey),
            KeyCode::PageDown => self.state.transition(TuiWidgetEvent::NextPageKey),
            KeyCode::Up => self.state.transition(TuiWidgetEvent::UpKey),
            KeyCode::Down => self.state.transition(TuiWidgetEvent::DownKey),
            KeyCode::Left => self.state.transition(TuiWidgetEvent::LeftKey),
            KeyCode::Right => self.state.transition(TuiWidgetEvent::RightKey),
            KeyCode::Char('h') => self.state.transition(TuiWidgetEvent::HideKey),
            KeyCode::Char('f') => self.state.transition(TuiWidgetEvent::FocusKey),
            KeyCode::Char('l') => return Ok(UIAction::ChangeView(UiPage::Voting)),
            _ => {}
        }
        return Ok(UIAction::Continue);
    }
}

fn help_spans<'a>(key: &'a str, description: &'a str) -> Vec<Span<'a>> {
    vec![
        Span::raw(key).style(Style::new().bold()),
        Span::raw(" - ").style(Style::new().gray()),
        Span::raw(description),
        Span::raw(" | ").style(Style::new().gray()),
    ]
}