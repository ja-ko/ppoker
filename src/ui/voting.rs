use std::cmp::Ordering;
use std::collections::HashMap;
use std::ops::{AddAssign, DerefMut};
use std::time::Instant;

use crossterm::event::{KeyCode, KeyEvent, KeyModifiers};
use ratatui::Frame;
use ratatui::prelude::*;
use ratatui::widgets::{Bar, BarChart, BarGroup, Cell, List, ListDirection, ListItem, ListState, Paragraph, Row, Table, Wrap};
use tui_big_text::{BigText, PixelSize};

use crate::app::{App, AppResult};
use crate::models::{GamePhase, LogLevel, LogSource, Player, UserType, Vote, VoteData};
use crate::ui::{colored_box_style, footer_entries, format_duration, Page, render_box, render_box_colored, render_confirmation_box, trim_name, UIAction, UiPage};

#[derive(Debug, PartialEq, Clone, Copy)]
pub enum InputMode {
    Menu,
    Vote,
    Name,
    Chat,
    RevealConfirm,
    ResetConfirm,
}

pub struct VotingPage {
    pub input_mode: InputMode,
    pub input_buffer: Option<String>,
    last_phase: GamePhase
}

impl Page for VotingPage {
    fn render(&mut self, app: &mut App, frame: &mut Frame) {
        app.has_updates = false;

        let chunks = Layout::default()
            .direction(Direction::Vertical)
            .constraints([
                Constraint::Length(3),
                Constraint::Fill(1),
                Constraint::Length(3)
            ])
            .split(frame.size());

        let header = chunks[0];
        let primary = chunks[1];
        let footer = chunks[2];

        let chunks = Layout::default()
            .direction(Direction::Horizontal)
            .constraints([
                Constraint::Percentage(30),
                Constraint::Min(26),
            ]).split(primary);

        let left_side = chunks[0];
        let right_side = chunks[1];

        let chunks = Layout::default()
            .direction(Direction::Vertical)
            .constraints([
                Constraint::Length(9),
                Constraint::Fill(1),
            ]).split(right_side);

        let vote_view = chunks[0];
        let log = chunks[1];

        if app.room.phase != self.last_phase {
            if self.input_mode != InputMode::Name {
                self.input_mode = InputMode::Menu;
            }
            self.last_phase = app.room.phase;
        }

        match app.room.phase {
            GamePhase::Revealed if app.history.len() > 0 => {
                let entry = app.history.as_slice().last().expect("Can't get last item of history.");
                render_own_vote(&entry.votes, entry.average, GamePhase::Revealed, &entry.own_vote, &entry.deck, vote_view, frame);
            }
            _ => {
                render_own_vote(&app.room.players, app.average_votes(), app.room.phase, &app.vote, &app.room.deck, vote_view, frame);
            }
        }
        self.render_log(app, log, frame);
        self.render_votes(app, left_side, frame);
        render_overview(app, header, frame);
        self.render_footer(app, footer, frame);
    }

    fn input(&mut self, app: &mut App, event: KeyEvent) -> AppResult<UIAction> {
        match &self.input_mode {
            InputMode::Menu => {
                match event.code {
                    KeyCode::Esc | KeyCode::Char('q') => {
                        return Ok(UIAction::Quit);
                    }
                    KeyCode::Char(c) if c.is_ascii_digit() => {
                        self.change_mode(InputMode::Vote, c.to_string(), app);
                    }
                    KeyCode::Char('-') => {
                        self.change_mode(InputMode::Vote, String::from("-"), app)
                    }
                    KeyCode::Char('v') => {
                        self.change_mode(InputMode::Vote, String::new(), app)
                    }
                    KeyCode::Char('c') if !event.modifiers.contains(KeyModifiers::CONTROL) => {
                        self.change_mode(InputMode::Chat, String::new(), app)
                    }
                    KeyCode::Char('n') => {
                        self.change_mode(InputMode::Name, app.name.clone(), app)
                    }
                    KeyCode::Char('l') => {
                        return Ok(UIAction::ChangeView(UiPage::Log));
                    }
                    KeyCode::Char('r') => {
                        if app.room.phase == GamePhase::Playing {
                            if app.room.players.iter().any(|p| p.user_type != UserType::Spectator && p.vote == Vote::Missing) {
                                self.input_mode = InputMode::RevealConfirm;
                            } else {
                                app.reveal()?;
                            }
                        } else {
                            self.input_mode = InputMode::ResetConfirm;
                        }
                    }
                    KeyCode::Char('h') => {
                        return Ok(UIAction::ChangeView(UiPage::History));
                    }
                    _ => {}
                }
            }
            InputMode::Vote | InputMode::Name | InputMode::Chat => {
                match event.code {
                    KeyCode::Esc => {
                        self.cancel_input();
                    }

                    KeyCode::Enter => {
                        self.confirm_input(app)?;
                    }

                    KeyCode::Backspace => {
                        if let Some(input_buffer) = &mut self.input_buffer {
                            input_buffer.pop();
                        }
                    }

                    KeyCode::Char(c) => {
                        if let Some(input_buffer) = &mut self.input_buffer {
                            input_buffer.push(c);
                        }
                    }
                    _ => {}
                }
            }
            InputMode::ResetConfirm => {
                match event.code {
                    KeyCode::Char('y') | KeyCode::Enter => {
                        app.restart()?;
                        self.input_mode = InputMode::Menu;
                    }
                    KeyCode::Char('n') | KeyCode::Esc => { self.input_mode = InputMode::Menu; }
                    KeyCode::Char('q') => { return Ok(UIAction::Quit); }
                    _ => {}
                }
            }
            InputMode::RevealConfirm => {
                match event.code {
                    KeyCode::Char('y') | KeyCode::Enter => {
                        app.reveal()?;
                        self.input_mode = InputMode::Menu;
                    }
                    KeyCode::Char('n') | KeyCode::Esc => { self.input_mode = InputMode::Menu; }
                    KeyCode::Char('q') => { return Ok(UIAction::Quit); }
                    _ => {}
                }
            }
        }
        Ok(UIAction::Continue)
    }

    fn pasted(&mut self, _app: &mut App, text: String) {
        match self.input_mode {
            InputMode::Chat | InputMode::Vote | InputMode::Name => {
                if let Some(input_buffer) = &mut self.input_buffer {
                    input_buffer.push_str(text.as_str());
                }
            }
            _ => {}
        }
    }
}


impl VotingPage {
    pub fn new() -> Self {
        Self {
            input_mode: InputMode::Menu,
            input_buffer: None,
            last_phase: GamePhase::Playing,
        }
    }

    pub fn change_mode(&mut self, mode: InputMode, default_text: String, app: &App) {
        if mode == InputMode::Vote && app.room.phase == GamePhase::Playing {
            self.start_input(mode, default_text)
        } else if mode == InputMode::Name || mode == InputMode::Chat {
            self.start_input(mode, default_text)
        }
    }

    fn start_input(&mut self, mode: InputMode, default: String) {
        self.input_mode = mode;
        self.input_buffer = Some(default);
    }

    pub fn confirm_input(&mut self, app: &mut App) -> AppResult<()> {
        let buffer = self.input_buffer.as_ref().map(|b| b.trim().replace('\n', ""));
        match self.input_mode {
            InputMode::Vote if app.room.phase == GamePhase::Playing => {
                if let Some(input_buffer) = &buffer {
                    let vote = input_buffer.clone();
                    app.vote(vote.as_str())?;
                }
                self.cancel_input();
            }
            InputMode::Name => {
                if let Some(input_buffer) = &buffer {
                    let name = input_buffer.clone();
                    app.rename(name)?;
                }
                self.cancel_input();
            }
            InputMode::Chat => {
                if let Some(input_buffer) = &buffer {
                    let name = input_buffer.clone();
                    app.chat(name)?;
                }
                self.cancel_input();
            }
            _ => {}
        }

        Ok(())
    }

    pub fn cancel_input(&mut self) {
        self.input_mode = InputMode::Menu;
        self.input_buffer = None;
    }

    fn render_votes(&mut self, app: &mut App, rect: Rect, frame: &mut Frame) {
        let rect = render_box_colored("Players", colored_box_style(app.room.phase), rect, frame);

        let mut longest_name: usize = 0;

        let mut players = app.room.players.clone();
        if app.room.phase == GamePhase::Revealed {
            fn vote_rank(vote: &Vote) -> i32 {
                match vote {
                    Vote::Missing => { 9999 }
                    Vote::Hidden => { 9999 }
                    Vote::Revealed(VoteData::Number(n)) => { *n as i32 }
                    Vote::Revealed(VoteData::Special(_)) => { 999 }
                }
            }
            players.sort_by(|p, p2| {
                let vote_order = vote_rank(&p.vote).cmp(&vote_rank(&p2.vote));
                if vote_order == Ordering::Equal {
                    p.name.cmp(&p2.name)
                } else {
                    vote_order
                }
            })
        } else {
            players.sort_by(|p, p2| {
                p.name.cmp(&p2.name)
            })
        }

        let rows: Vec<Row> = players.iter().map(|player| {
            let player_color = if player.is_you {
                Style::new().green()
            } else {
                Style::new()
            };
            let name = crate::ui::voting::trim_name(&player.name);
            if name.len() > longest_name {
                longest_name = name.len()
            }

            Row::new(vec![
                Cell::from(Span::styled(name, player_color)),
                Cell::from(format_vote(&player.vote, &app.vote)),
            ])
        }).collect();

        let table = Table::new(rows, [Constraint::Length(longest_name as u16), Constraint::Fill(1)])
            .column_spacing(3)
            .header(
                Row::new(vec!["Name", "Vote"])
                    .style(Style::new().bold())
                    .bottom_margin(1)
            );

        frame.render_widget(table, rect);
    }

    fn render_log(&mut self, app: &mut App, rect: Rect, frame: &mut Frame) {
        let rect = render_box_colored("Log", colored_box_style(app.room.phase), rect, frame);

        let entries: Vec<ListItem> = app.log.iter().map(|logentry| {
            let color = match logentry.level {
                LogLevel::Chat => { Style::new().light_blue() }
                LogLevel::Info => {
                    if logentry.source == LogSource::Server {
                        Style::new()
                    } else {
                        Style::new().yellow()
                    }
                }
                LogLevel::Error => { Style::new().red() }
            };
            let prefix = match logentry.level {
                LogLevel::Chat => { String::from("") }
                _ => {
                    format!("[{:?}]: ", logentry.source)
                }
            };
            ListItem::new(format!("{}{}", prefix, logentry.message)).style(color)
        }).collect();

        let mut state = ListState::default().with_offset(entries.len().saturating_sub(rect.height as usize));
        let list = List::new(entries)
            .direction(ListDirection::TopToBottom);

        frame.render_stateful_widget(list, rect, &mut state);
    }

    fn render_footer(&mut self, app: &mut App, rect: Rect, frame: &mut Frame) {
        match &self.input_mode {
            InputMode::Vote => {
                let layout = Layout::default()
                    .direction(Direction::Horizontal)
                    .constraints([Constraint::Length(20), Constraint::Fill(1)])
                    .split(rect);

                self.render_text_input("Vote", layout[0], frame);
                let mut spans: Vec<Span> = app.room.deck.iter().flat_map(|item| {
                    vec![
                        Span::raw(" "),
                        Span::raw(item.clone()),
                        Span::raw(" |"),
                    ]
                }).collect();

                spans.insert(0, Span::raw("   Possible values:"));
                spans.remove(spans.len() - 1);


                let possible_values = Paragraph::new(vec![Line::from(""), Line::from(spans)]).style(Style::new().gray());
                frame.render_widget(possible_values, layout[1]);
            }
            InputMode::Name => {
                self.render_text_input("Rename", rect, frame);
            }
            InputMode::Chat => {
                self.render_text_input("Chat", rect, frame);
            }
            InputMode::RevealConfirm => {
                render_confirmation_box("Not everyone has voted yet. Confirm you want to reveal the cards?", rect, frame);
            }
            InputMode::ResetConfirm => {
                render_confirmation_box("Confirm you want to reset the game?", rect, frame);
            }
            InputMode::Menu => {
                let entries = if app.room.phase == GamePhase::Playing {
                    vec!["Vote", "Reveal", "Name change", "Chat", "Quit"]
                } else {
                    vec!["Restart", "Name change", "Chat", "Quit"]
                };

                frame.render_widget(footer_entries(entries), rect);
            }
        }
    }

    fn render_text_input(&mut self, title: &str, rect: Rect, frame: &mut Frame) {
        let rect = render_box(title, rect, frame);
        let buffer = self.input_buffer.as_ref().map_or("", |buffer| buffer.as_str());
        let text_buffer = Paragraph::new(buffer);
        frame.render_widget(text_buffer, rect);
        frame.set_cursor(
            rect.x + buffer.len() as u16,
            rect.y,
        );
    }
}

pub(super) fn render_own_vote(players: &Vec<Player>, average_vote: f32, phase: GamePhase, own_vote: &Option<VoteData>, deck: &Vec<String>, rect: Rect, frame: &mut Frame) {
    let constraints = if phase == GamePhase::Revealed {
        [
            Constraint::Length(26),
            Constraint::Length((deck.len() * 3) as u16),
            Constraint::Length(34),
        ]
    } else {
        [
            Constraint::Length(26),
            Constraint::Fill(1),
            Constraint::Fill(1),
        ]
    };
    let chunks = Layout::default()
        .direction(Direction::Horizontal)
        .constraints(constraints)
        .split(rect);
    let small_box = chunks[0];
    let bar_chart = chunks[1];
    let average = chunks[2];

    if phase == GamePhase::Revealed {
        let inner = render_box_colored("Vote distribution", colored_box_style(phase), bar_chart, frame);

        let mut cards = HashMap::new();
        for player in players {
            let card = format!("{}", player.vote);
            cards.entry(card).or_insert(0).deref_mut().add_assign(1);
        }

        let cards: Vec<_> = deck.iter().map(|card| {
            Bar::default()
                .text_value(card.clone())
                .value(*cards.get(card).unwrap_or(&0))
        }).collect();

        let chart = BarChart::default()
            .bar_width(2)
            .bar_gap(1)
            .data(BarGroup::default().bars(cards.as_slice()));

        frame.render_widget(chart, inner);

        let inner = render_box_colored("Average vote", colored_box_style(phase), average, frame);
        let text = BigText::builder()
            .pixel_size(PixelSize::Full)
            .style(Style::new().light_blue())
            .alignment(Alignment::Center)
            .lines(vec![format!("{:.1}", average_vote).into()])
            .build().expect("Failed to build Text widget");
        frame.render_widget(text, inner);
    }

    let inner = render_box_colored("Your vote", colored_box_style(phase), small_box, frame);

    let (color, text) = if let Some(vote) = &own_vote {
        (Style::new().green(), vote.to_string())
    } else {
        (Style::new().red(), "-".to_owned())
    };

    let text = BigText::builder()
        .pixel_size(PixelSize::Full)
        .style(color)
        .alignment(Alignment::Center)
        .lines(vec![text.into()])
        .build().expect("Failed to build text widget");
    frame.render_widget(text, inner);
}

pub(super) fn render_overview(app: &mut App, rect: Rect, frame: &mut Frame) {
    let rect = render_box("Overview", rect, frame);

    let name = trim_name(app.name.as_str());
    let state_color = if app.room.phase == GamePhase::Playing {
        Style::new().yellow()
    } else {
        Style::new().light_blue()
    };

    let duration = if app.room.phase == GamePhase::Revealed && app.history.len() > 0 {
        format_duration(&app.history[app.history.len() - 1].length)
    } else {
        format_duration(&(Instant::now() - app.round_start))
    };

    let text = Line::from(vec![
        Span::raw("Name: "),
        Span::raw(name).bold(),
        Span::raw(" | Room: "),
        Span::raw(app.room.name.as_str()).bold(),
        Span::raw(" | Server: "),
        Span::raw(app.config.server.as_str()).bold(),
        Span::raw(" | State: "),
        Span::raw(format!("{}", app.room.phase)).style(state_color.bold()),
        Span::raw(" | Round: "),
        Span::raw(app.round_number.to_string()).bold(),
        Span::raw(format!(" ({})", duration)),
    ]);

    let paragraph = Paragraph::new(text)
        .alignment(Alignment::Left)
        .wrap(Wrap { trim: true });
    frame.render_widget(paragraph, rect);
}

pub fn format_vote(vote: &Vote, own_vote: &Option<VoteData>) -> Span<'static> {
    match vote {
        Vote::Missing => { Span::raw("-").style(Style::new().red()) }
        Vote::Hidden => { Span::raw("#").style(Style::new().green()) }
        Vote::Revealed(data) => {
            match data {
                VoteData::Number(n) => {
                    let color = if let Some(VoteData::Number(n2)) = own_vote {
                        if *n2 == *n {
                            Style::new().green()
                        } else if *n2 < *n {
                            Style::new().light_blue()
                        } else {
                            Style::new().yellow()
                        }
                    } else {
                        Style::new()
                    };
                    Span::raw(n.to_string()).style(color)
                }
                VoteData::Special(t) => {
                    if t.trim().is_empty() {
                        Span::raw("-").style(Style::new().red())
                    } else {
                        Span::raw(t.clone())
                    }
                }
            }
        }
    }
}
