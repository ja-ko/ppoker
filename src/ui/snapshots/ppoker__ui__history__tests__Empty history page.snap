---
source: src/ui/history.rs
expression: terminal.backend()
---
"╭Overview──────────────────────────────────────────────────────────────────────╮"
"│Name: Test User | Room: Planning Room | Server: wss://mocked | State: Playing │"
"╰──────────────────────────────────────────────────────────────────────────────╯"
"╭History───────────────────────╮                                                "
"│Round    Average     Duration │                                                "
"│                              │                                                "
"│                              │                                                "
"│                              │                                                "
"│                              │                                                "
"│                              │                                                "
"│                              │                                                "
"│                              │                                                "
"│                              │                                                "
"│                              │                                                "
"│                              │                                                "
"│                              │                                                "
"╰──────────────────────────────╯                                                "
"                                                                                "
" Vote | ↑ | ↓ | Quit                                                            "
"                                                                                "
